core/src/main/scala/org/apache/spark/resource/ResourceAllocator.scala

```diff
@@ -86,8 +103,8 @@ trait ResourceAllocator {
           s"address $address doesn't exist.")
       }
       val isAvailable = addressAvailabilityMap(address)
-      if (!isAvailable) {
-        addressAvailabilityMap(address) = true
+      if (isAvailable < slotsPerAddress) {
+        addressAvailabilityMap(address) = addressAvailabilityMap(address) + 1
```

Comment:
Does it worth add `synchronized` to the `addressAvailabilityMap` to ensure thread safe?

Reply:
`ResourceAllocator` _should_ only used from a single thread (it is documented that way).

For its public interface, here is the access pattern:

- `CoarseGrainedSchedulerBackend` and `Master` event loop are calling into `acquire` and `release`. They are also using `availableAddrs`, and are calling `toBuffer` on it before handing off that state. This is a single thread pattern.

- A class extending `ResourceAllocator` (`WorkerResourceInfo`), has some potential issues: The `WorkerInfo` class is calling into `availableAddrs` and `assignedAddrs` but those calls appear to be coming from the UI (looking at the `resourcesInfo*` functions), e.g. `JsonProtocol` and `MasterPage` call this. There are other calls to the `resourceInfo*` functions, but those are from the event loop. Is this what you were worried about @jiangxb1987?

Note this issue is not directly related to this PR, as this PR didn't change that access pattern, or the datastructure. That said, I do believe that a request from the UI could result in at least an exception (if not worse). Let me know and I can handle here, or open another PR that would address it.

Reply:
yeah I think the only issue here are with UI accessing, and I probably missed this in the reviews for the standalone support. I was thinking it was just read but its doing a flatMap in there.
But I do think we should protect that. That said it is a separate issue not introduced by this so I think we should file a separate jira for it, @jiangxb1987 are you ok with that?

Reply:
sure, it worth a separate jira.

Reply:
https://issues.apache.org/jira/browse/SPARK-29780